latch_cli/services/cp/upload.py

```diff
@@ -0,0 +1,366 @@
+import math
+import mimetypes
+import os
+import time
+from concurrent.futures import Future, ProcessPoolExecutor, wait
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Dict, List, Optional, TypedDict
+
+import click
+
+from latch_cli import tinyrequests
+from latch_cli.config.latch import config as latch_config
+from latch_cli.constants import latch_constants, units
+from latch_cli.services.cp.config import CPConfig, Progress
+from latch_cli.services.cp.ldata_utils import LDataNodeType, get_node_data
+from latch_cli.services.cp.path_utils import normalize_path, remote_joinpath
+from latch_cli.services.cp.progress import ProgressBarManager, ProgressBars
+from latch_cli.services.cp.utils import get_auth_header, human_readable_time, pluralize
+from latch_cli.utils import with_si_suffix
+
+
+class EmptyUploadData(TypedDict):
+    version_id: str
+
+
+class StartUploadData(TypedDict):
+    upload_id: str
+    urls: List[str]
+
+
+def upload(
+    src: str,  # pathlib.Path strips trailing slashes but we want to keep them here as they determine cp behavior
+    dest: str,
+    config: CPConfig,
+):
+    src_path = Path(src)
+    if not src_path.exists():
+        raise ValueError(f"Could not find {src_path}.")
+
+    normalized = normalize_path(dest)
+    dest_data = get_node_data(normalized, allow_resolve_to_parent=True)
+
+    dest_exists = not dest_data.is_parent
+    dest_is_dir = dest_data.type in {
+        LDataNodeType.account_root,
+        LDataNodeType.mount,
+        LDataNodeType.dir,
+    }
+
+    if config.progress == Progress.none:
+        num_bars = 0
+        show_total_progress = False
+    elif not src_path.is_dir():
+        num_bars = 1
+        show_total_progress = False
+    elif config.progress == Progress.total:
+        num_bars = 0
+        show_total_progress = True
+    else:
+        num_bars = config.max_concurrent_files
+        show_total_progress = True
+
+    with (
+        ProcessPoolExecutor() as executor,
+        ProgressBarManager() as manager,
+    ):
+        if src_path.is_dir():
+            if not dest_is_dir:
+                raise ValueError(f"{normalized} is not a directory.")
+
+            if dest_exists and not src.endswith("/"):
```

Comment:
so this uploads `a/b/hello` to `dest` as `dest/hello/*`
+ `a/b/hello/` to `dest` as `dest/*`?

this seems awful. is that's what unix `cp` does? i fucking hate unix
add a todo to come up with a better scheme

Reply:
yessir that is unix cp behavior

Reply:
todo pls